### PR TITLE
Migrate to vite.js and Typescript (fully typed)

### LINKED ISSUE #133: Move to TypeScript

### PR DESCRIPTION
**Pull request by spaenleh**
This PR adds vite as the bundler (instead of Create React App) and moves all files to typescript.

The code organisation is inspired by **bullet proof react** with the `modules` folder and then repartition by "feature". 

Imports are absolute with path aliases:
- `@/` points to `/src`

closes #133

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
```
Comment (pyphilia) [inline index.html:1]: So now we need index.html at the root?

File: package.json [hunk 2]
```diff
@@ -10,49 +10,10 @@
...
-    "react-toastify": "9.0.3",
-    "uuid": "9.0.0"
-  },
   "scripts": {
-    "pre-commit": "yarn prettier:check && yarn lint",
-    "post-commit": "git status",
-    "start": "env-cmd -f ./.env.local react-scripts start",
-    "start:test": "env-cmd -f ./.env.test react-scripts start",
-    "start:prod": "env-cmd -f ./.env.production react-scripts start",
-    "start:ci": "react-scripts -r @cypress/instrument-cra start",
```
Comment (pyphilia) [inline package.json:51]: What about coverage?

File: package.json [hunk 3]
```diff
@@ -64,57 +25,66 @@
...
-      "last 1 chrome version",
-      "last 1 firefox version",
-      "last 1 safari version"
-    ]
+  "dependencies": {
+    "@emotion/react": "11.10.6",
+    "@emotion/styled": "11.10.6",
+    "@graasp/chatbox": "github:graasp/graasp-chatbox#main",
+    "@graasp/query-client": "github:graasp/graasp-query-client#main",
+    "@graasp/sdk": "github:graasp/graasp-sdk#92-add-build-utilities-for-extra",
```
Comment (pyphilia) [inline package.json:33]: To update ;)

File: package.json [hunk 4]
```diff
@@ -64,57 +25,66 @@
...
+    "typescript": "^4.9.5",
+    "vite": "^4.1.0",
+    "vite-plugin-checker": "^0.5.5",
+    "vite-plugin-istanbul": "4.0.1"
   },
-  "packageManager": "yarn@3.4.1",
   "resolutions": {
-    "react-error-overlay": "6.0.11"
-  }
+    "@graasp/sdk": "github:graasp/graasp-sdk#92-add-build-utilities-for-extra",
```
Comment (pyphilia) [inline package.json:85]: to remove ;)

File: src/config/constants.ts [hunk 5]
```diff
@@ -61,30 +32,36 @@ export const DESCRIPTION_MAX_LENGTH = 130;
...
+
+export const AVATAR_ICON_HEIGHT = 30;
+
+export const ThumbnailSizeVariants = {
+  SMALL: 'small',
+  MEDIUM: 'medium',
+  LARGE: 'large',
+} as const;
+export type ThumbnailSizeVariantsType =
+  `${(typeof ThumbnailSizeVariants)[keyof typeof ThumbnailSizeVariants]}`;
```
Comment (pyphilia) [inline src/config/constants.ts:67]: We can add them to sdk.

File: src/config/env.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+export const ENV = {
+  PRODUCTION: 'production',
+  DEVELOPMENT: 'development',
+  TEST: 'test',
+};
```
Comment (pyphilia) [inline src/config/env.ts:5]: Should we add this to sdk too 🤔

File: src/modules/item/ItemScreen.tsx [hunk 7]
```diff
@@ -20,25 +20,17 @@ const ItemScreen = () => (
...
+  const { mutate: itemLoginSignIn } = usePostItemLogin();
   const { rootId } = useParams();
 
   const ForbiddenContent = <ItemForbiddenScreen />;
 
   const Component = ItemLoginAuthorization({
     signIn: itemLoginSignIn,
     signOut,
-    itemId: rootId,
+    itemId: rootId || '',
```
Comment (pyphilia) [inline src/modules/item/ItemScreen.tsx:33]: isn't null better? If it's too avoid changing ui you can add a comment with the better solution maybe.

**Review by pyphilia — CHANGES_REQUESTED** [review]
Nice work, can't wait to see the improvement while working on player 😮

File: .github/workflows/cintegration-s3-caller.yml [hunk 8]
```diff
@@ -1,52 +1,69 @@
-name: Deploy to development environment
+name: Deploy app to dev
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply .github/workflows/cintegration-s3-caller.yml:1]: 🤷 I can change back

File: .github/workflows/deploy-dev-template.yml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+name: Deploy app to dev
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply .github/workflows/deploy-dev-template.yml:1]: I wanted to change the name of the workflow but I guess this should be done in another PR.

File: .yarn/sdks/eslint/bin/eslint.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env node
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply .yarn/sdks/eslint/bin/eslint.js:1]: This is some leftovers from when I tried to use yarn's Plug-n-play feature. I will remove them.

File: cypress/e2e/ws.cy.ts [hunk 11]
```diff
@@ -59,9 +61,9 @@ describe('Websocket interactions', () => {
       });
   });
 
-  it('Displays remove child update', () => {
+  it.skip('Displays remove child update', () => {
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply cypress/e2e/ws.cy.ts:64]: 🧨 💣 🔪 💀 Good catch !

File: index.html (hunk 1 above)
Comment (spaenleh) [reply index.html:1]: Yes this is a specificity of Vite. The entry point is `index.html` and it should be at the root.

Comment (spaenleh) [reply index.html:1]: See [index.html and project root in `vite.js` docs](https://vitejs.dev/guide/#index-html-and-project-root).

File: package.json (hunk 2 above)
Comment (spaenleh) [reply package.json:51]: Coverage with cypress in vite is handled differently than with CRA.
Actually, it is CRA that handled it differently (but this is another topic).

In CRA:
- use the "require" option of `react-scripts` to require packages execution before starting the app.
  - this was needed to instrument the app before it was run

In Vite:
- use `vite-plugin-istanbul` in vite config to instrument app code with istanbul. No Babel transforms needed.

What does not change:
- register the `@cypress/coverage/task` in the node events sections of the cypress config.

Hope it makes sense, feel free to follow up on any questions.

Comment (spaenleh) [reply package.json:51]: You can check the coverage results in [the latest run](https://github.com/graasp/graasp-player/actions/runs/4355445034/jobs/7612073787) under the `coverage report` step.

File: package.json (hunk 3 above)
Comment (spaenleh) [reply package.json:33]: Need a review on https://github.com/graasp/graasp-sdk/pull/93

File: src/config/env.ts (hunk 6 above)
Comment (spaenleh) [reply src/config/env.ts:5]: With vite we can just check for `import.meta.env.DEV` where we do `NODE_ENV === ENV.DEVELOPMENT`. So we should be able to remove the `ENV`. And use the `mode` from vite instead.

**Review by swouf — COMMENTED** [review]
Nice!

I guess that all files in [.yarn/sdks/](https://github.com/graasp/graasp-player/pull/204/files#diff-c24bac507d46463d067cf08a70a0cf3ccead1a04d36636d8a6c0ef6f65398b16) are automatically generated so I didn't review them (I guess they can be trusted).

File: .github/workflows/cintegration-s3-caller.yml [hunk 12]
```diff
@@ -1,52 +1,69 @@
...
+          VITE_GA_MEASUREMENT_ID: ${{ secrets.REACT_APP_GA_MEASUREMENT_ID_DEV }}
+          VITE_GRAASP_BUILDER_HOST: ${{ secrets.BUILDER_CLIENT_HOST_DEV }}
+          VITE_GRAASP_ANALYTICS_HOST: 'not-used'
+          VITE_GRAASP_LIBRARY_HOST: ${{ secrets.EXPLORER_CLIENT_HOST_DEV }}
+          VITE_H5P_INTEGRATION_URL: ${{ secrets.H5P_INTEGRATION_URL_DEV }}
+          VITE_HIDDEN_ITEM_TAG_ID: ${{ secrets.REACT_APP_HIDDEN_ITEM_TAG_ID_DEV }}
+          VITE_PORT: ${{ secrets.PORT }}
+          VITE_SENTRY_DSN: ${{ secrets.REACT_APP_SENTRY_DSN }}
+          VITE_SHOW_NOTIFICATIONS: ${{ secrets.REACT_APP_SHOW_NOTIFICATIONS }}
+          VITE_GRAASP_ASSETS_URL: ${{ secrets.REACT_APP_GRAASP_ASSETS_URL_DEV }}
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply .github/workflows/cintegration-s3-caller.yml:54]: Yes then environnement variables have changed.
In CRA env vars must be prefixed with `REACT_APP_` and in vite.js they need to be prefixed with `VITE_`.
Only variables with the right prefix are loaded by cra/vite. 

Not all of these env variables are secrets, so the next move will be to use Github environnements to store variables and secrets.

File: .github/workflows/deploy-dev-template.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (spaenleh) [reply .github/workflows/deploy-dev-template.yml:1]: There is no difference, I need to remove this file.

File: .github/workflows/deploy-dev-template.yml [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+          # Allows to increase Node's max heap size
+          # NODE_OPTIONS: '--max_old_space_size=8192'
+          # Prevent react to generate source maps
+          # GENERATE_SOURCEMAP: "false"
+
+          REACT_APP_GRAASP_DOMAIN: ${{ inputs.graasp-domain }}
+          REACT_APP_GRAASP_APP_ID: ${{ vars.app-id }}
+          REACT_APP_SENTRY_DSN: ${{ vars.sentry-dsn }}
+          REACT_APP_MOCK_API: "false"
+          REACT_APP_VERSION: ${{ inputs.tag }}
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply .github/workflows/deploy-dev-template.yml:46]: See above, I will remove this file.

File: src/config/constants.ts (hunk 5 above)
Comment (spaenleh) [reply src/config/constants.ts:67]: Done !

**Review by pyphilia — APPROVED** [review]
🐪

**Pull request merged**